tests/models/test_base_async_operator.py

```diff
@@ -0,0 +1,288 @@
+
+        async_op = DummyAsyncOperator(
+            task_id=ASYNC_OP,
+            return_value=return_value,
+            resource_id=resource_id,
+            dag=self.dag,
+            **kwargs
+        )
+
+        dummy_op = DummyOperator(
+            task_id=DUMMY_OP,
+            dag=self.dag
+        )
+        dummy_op.set_upstream(async_op)
+        return async_op
+
+    @classmethod
+    def _run(cls, task):
+        task.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE, ignore_ti_state=True)
+
+    def test_ok(self):
+        """ Test normal behavior"""
+        async_op = self._make_async_op(True)
+        dr = self._make_dag_run()
+
+        self._run(async_op)
+        tis = dr.get_task_instances()
+        self.assertEqual(len(tis), 2)
+        for ti in tis:
+            if ti.task_id == ASYNC_OP:
+                self.assertEqual(ti.state, State.SUCCESS)
+            if ti.task_id == DUMMY_OP:
+                self.assertEqual(ti.state, State.NONE)
+
+    def test_poke_fail(self):
+        """ Test failure in poke"""
+        async_op = self._make_async_op(False)
+        dr = self._make_dag_run()
+
+        with self.assertRaises(AirflowSensorTimeout):
+            self._run(async_op)
+        tis = dr.get_task_instances()
+        self.assertEqual(len(tis), 2)
+        for ti in tis:
+            if ti.task_id == ASYNC_OP:
+                self.assertEqual(ti.state, State.FAILED)
+            if ti.task_id == DUMMY_OP:
+                self.assertEqual(ti.state, State.NONE)
+
+    @parameterized.expand(ALL_ID_TYPES)
+    def test_set_get_external_resource_id(self, resource_id):
+        """ test resource id mechanism """
+        async_op = self._make_async_op(
+            return_value=None,
+            poke_interval=10,
+            timeout=25)
+
+        context = TaskInstance(task=async_op,
+                               execution_date=DEFAULT_DATE).get_template_context()
+        async_op.set_external_resource_id(context, resource_id)
+        self.assertEqual(resource_id, async_op.get_external_resource_id(context))
+
+    def test_xcom(self):
+        """test xcom is set w/ job id. """
+        async_op = self._make_async_op(
+            return_value=None,
+            poke_interval=10,
+            timeout=25)
+        async_op.process_result = Mock()
+        async_op.poke = Mock(side_effect=[True])
+
+        dr = self._make_dag_run()
+
+        date1 = timezone.utcnow()
+        with freeze_time(date1):
+            self._run(async_op)
+        tis = dr.get_task_instances()
+
+        # Check that XCom was set to job_id.
+        for ti in tis:
+            if ti.task_id == ASYNC_OP:
+                resource_id = ti.xcom_pull(task_ids=ASYNC_OP,
+                                           key=XCOM_EXTERNAL_RESOURCE_ID_KEY)
+                self.assertIsNotNone(resource_id)
+                self.assertTrue(resource_id.startswith('job_id'))
+
+    def test_ok_with_reschedule(self):
```

Comment:
@dstandish I'll try to improve this unit test until it reproduces your issue w/ XCom clearing.

Reply:
The PR has been merged; https://github.com/apache/airflow/pull/6370

Reply:
Just pointing out here that #6370 was reverted ([link](https://gitbox.apache.org/repos/asf?p=airflow.git;h=4132f3b)). Are you intending to continue work on a variation of that, @Fokko?

Reply:
Yes, but I need more time to get this to work.

Reply:
@jaketf @JonnyIncognito I'm getting a lot of pushback on idempotency, which is a fair point. Keeping state makes the function inpure by definition. How about adding labels to the services that we launch on GCP. This label should contain a tuple of `(composer_instance, dag_id, task_id, execution_date)` to make it unique. In this case, we could just query the state using the GCP API's. WDYT?

Reply:
@Fokko are you considering abandoning the persisting of TI state on metastore?  So you persist state externally?  I think think this TI state persistence is crucial feature and must be supported.  XCom can support it with appropriate changes.  We just have to figure out what those are.  Another option is just let writer of operator choose when xcom is cleared by providing suitable methods to TaskInstance.

Reply:
I would like to investigate other options. I'm not ruling anything out, but I think about it. If we could avoid storing state in the metastore, then it would make life simpler, and I like simple.

Reply:
@Fokko the ability to retain state is super important for a rescheduled task, I agree with @dstandish on this. These are mostly used where external long-lived resources are created. For sensors we've dodged the issue because they tend to use XCom to fetch the state from a prior task. But if we're now going further and having multi-phased / composed tasks, there needs to be a general Airflow mechanism for that state rather than hacking around it with e.g. cloud-provider specific things like labels.

> Keeping state makes the function impure by definition.

Depends on the scope of the idempotency. I can see why this argument could be used for existing operators that are supposed to do the same thing for every call of `execute()`. However, for ones with multiple phases they should be idempotent by the time the task succeeds, across multiple calls to `execute()`, i.e. the scope is a bit wider.

How about we only clear the state when needed? Options:

1. Delegate clearing to the task (default implementation always clears). If you look at my proposal in other thread ([link](https://github.com/apache/airflow/pull/6210#discussion_r339016148)) the task could clear all state on the initial phase but not for subsequent phases / reschedules.

1. Clear state only if there was no prior transition to `State.UP_FOR_RESCHEDULE`. I've seen code that detects this but it'd add a cost for every user ([link](https://github.com/apache/airflow/blob/master/airflow/sensors/base_sensor_operator.py#L104)).

1. Add some concept of a reschedule counter to the task instance and only clear all state on the first execution. If it were more along the lines of a phase counter, it could later be used for tasks that have more than two phases to figure out which code path to follow on each `execute()`.

Reply:
So the problem is with the `ti.xcom_pull(include_prior_dates=true)`. If you run this one on the first poll, then it might be different than with subsequent polls. Personally I would be fine with the first one option and having some specific key for storing the state.